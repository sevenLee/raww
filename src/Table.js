import React, { Component } from 'react'
import { Table } from 'react-bootstrap'

export class RepoTable extends Component{
    makeTitle(colMapElm, prop) {
        const  {config: {sortCol, descending}, sortColState, descendingState, onSorting} = this.props
        const sortActive = (sortCol === prop)
        const title = () => {
            if(colMapElm.headerContent && typeof colMapElm.headerContent === 'function') {
                return colMapElm.headerContent();
            }

            return colMapElm.headerContent;
        }


        if (!colMapElm.sorting) {
            return <span>{ title() }</span>
        }

        return (
            <span
                className={`sorting-title${sortActive? ' active' : ''}`}
                onClick={(e) => {
                    colMapElm.sorting.onClick(e, prop, sortColState, descendingState, onSorting)
                }}
            >
                { title() }
                {
                    sortActive?
                        descending?
                            <span dangerouslySetInnerHTML={{ __html: '&nbsp;\u25BC' }} /> :
                            <span dangerouslySetInnerHTML={{ __html: '&nbsp;\u25B2' }} />
                        :
                        <span dangerouslySetInnerHTML={{ __html: '&nbsp;\u25BD' }} />
                }
            </span>
        )

    }

    makeFilter(colMapElm, prop) {
        const {sortColState, descendingState, onFilting} = this.props
        return (
            (colMapElm.filter)?
                <div className="filter">
                    <input
                        onChange={(e) => colMapElm.filter.onChange(e, prop, sortColState, descendingState, onFilting)}
                    />
                </div>
                :
                null
        )
    }

    makeTHNode(prop, colMapElm) {
        const { config: {hasIndex} } = this.props
        let inlineStyle = (colMapElm.thInlineStyle) ? colMapElm.thInlineStyle : {}
        let classStyle = colMapElm.align || 'text-right'

        if(hasIndex && prop === 'firstColIndex') {
            inlineStyle = Object.assign(inlineStyle, {
                width: 45
            })
        }

        return (
            <th className={(colMapElm.sorting)?`${classStyle} sorting`:classStyle} style={inlineStyle} key={colMapElm.index} >
                {this.makeTitle(colMapElm, prop)}
                {this.makeFilter(colMapElm, prop)}
            </th>
        )
    }

    renderHeader() {
        const { colMap, config } = this.props
        const visibleHeaders = []

        for(let prop in colMap) {
            const colMapElm = colMap[prop]
            visibleHeaders[colMapElm.index] = this.makeTHNode(prop, colMapElm, config)
        }

        return visibleHeaders
    }

    makeCell(prop, row, colMapElm) {
        let cellVal = row[prop]

        switch(colMapElm.style){
            case 'custom':
            {
                if(colMapElm.makeCell && typeof colMapElm.makeCell === 'function'){
                    return colMapElm.makeCell(cellVal, row)
                }else if(colMapElm.makeCell && typeof colMapElm.makeCell[0] === 'function'){
                    let argumentList = colMapElm.makeCell.filter((elm, index) => index !== 0)
                    argumentList.unshift(row)
                    argumentList.unshift(cellVal)
                    return colMapElm.makeCell[0](...argumentList)
                }else {
                    return cellVal
                }
            }
            default:
            {
                return (
                    (colMapElm.onCellClick) ?
                        <a onClick={() => colMapElm.onCellClick(cellVal)}>{cellVal}</a>
                        :
                        cellVal
                )
            }
        }
    }

    renderRows() {
        const { colMap, config: { hasIndex, trClass, trOnClick, pageNum, pageSize=20 }, datas } = this.props
        const visibleRows = []
        const cellNum = Object.keys(colMap).length

        datas.forEach((row, index) => {
            const visibleCells = []

            // reset
            for(let i=0; i<cellNum; i++){
                visibleCells[i] = (
                    <td key={i} className={'text-right'} />
                )
            }

            const makeTRClass = () => {
                return (typeof trClass === 'function') ? trClass() : trClass
            }

            let trSingleClass = (row.customRepoRow)? makeTRClass() :''

            for(let prop in row) {
                const colMapElm = colMap[prop]
                if(colMapElm) {
                    let inlineStyle = (colMapElm.tdInlineStyle) ? colMapElm.tdInlineStyle : {}
                    let classStyle = colMapElm.align || 'text-right'

                    if(colMapElm.replaceTrClass) {
                        trSingleClass = colMapElm.replaceTrClass(trSingleClass, row[prop], row)
                    }

                    visibleCells[colMapElm.index] = (
                        <td key={colMapElm.index}
                            className={classStyle}
                            style={inlineStyle}>
                            {
                                this.makeCell(prop, row, colMapElm)
                            }
                        </td>
                    )
                }
            }

            if(hasIndex && !row.customRepoRow) {
                let classStyle = ''
                if(colMap.firstColIndex && colMap.firstColIndex.align){
                    classStyle = colMap.firstColIndex.align || 'text-right'
                }

                index = index+1
                if(pageNum){
                    index = index + (pageNum-1)*pageSize
                }

                visibleCells[0] = (
                    <td className={classStyle + " text-index"} key={0} >{index}</td>
                )
            }

            visibleRows[index] = (
                <tr
                    key={index}
                    onClick={(e) => trOnClick && trOnClick(e, row)}
                    className={trSingleClass}
                >
                    {visibleCells}
                </tr>
            )
        })
        return visibleRows
    }

    render() {
        const { config, config: {addClass}, descriptionComp } = this.props
        const defaultProps = {
            config: {
                responsive: true
            }
        }

        const finalConfig = Object.assign({}, defaultProps.config, config)

        let baseClass = ['bo-table']
        const tableClassList = (addClass) ? baseClass.concat(addClass) : baseClass
        let tableClass = tableClassList.join(' ')

        if(config.replaceClass) {
            tableClass = config.replaceClass
        }

        return (
            <div className="table-wrapper">
                {
                    descriptionComp &&
                    <div className="table-description">{descriptionComp}</div>
                }

                <Table
                    responsive={finalConfig.responsive}
                    className={tableClass}
                >
                    <thead>
                    <tr>
                        {this.renderHeader()}
                    </tr>
                    </thead>
                    <tbody>
                    {this.renderRows()}
                    </tbody>
                    {this.props.children}
                </Table>
            </div>
        )
    }
}

RepoTable.defaultProps = {
    config: {},
    colMap: {}
}

export default RepoTable
