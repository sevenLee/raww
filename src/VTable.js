import React, { Component } from 'react'
import { Table } from 'react-bootstrap'

class VTable extends Component{

    makeCell(prop, data, colMapElm) {
        switch(colMapElm.style){
            case 'custom':
            {
                if(colMapElm.makeCell && typeof colMapElm.makeCell === 'function'){
                    return colMapElm.makeCell(data[prop], data)
                }else if(colMapElm.makeCell && typeof colMapElm.makeCell[0] === 'function'){
                    let argumentList = colMapElm.makeCell.filter((elm, index) => index !== 0)
                    argumentList.unshift(data)
                    argumentList.unshift(data[prop])
                    return colMapElm.makeCell[0](...argumentList)
                }else {
                    return data[prop]
                }
            }
            default:
            {
                return data[prop]
            }
        }
    }

    renderRows() {
        const { colMap, data } = this.props
        const visibleRows = []

        for(let prop in data) {
            const colMapElm = colMap[prop]
            const title = () => {
                if(colMapElm.headerContent && typeof colMapElm.headerContent === 'function') {
                    return colMapElm.headerContent();
                }

                return colMapElm.headerContent;
            }
            if(colMapElm) {
                visibleRows[colMapElm.index] = (
                    <tr key={colMapElm.index}>
                        <td key="1" className="head">{title()}</td>
                        <td key="2" className="content">{this.makeCell(prop, data, colMapElm)}</td>
                    </tr>
                )
            }
        }

        return visibleRows
    }

    render() {
        const {config} = this.props

        let baseClass = ['vertical-table']
        const tableClassList = (config && config.addClass) ? baseClass.concat(config.addClass) : baseClass
        let tableClass = tableClassList.join(' ')

        if(config.replaceClass) {
            tableClass = config.replaceClass
        }


        return (
            <Table responsive className={tableClass}>
                <tbody>
                {this.renderRows()}
                </tbody>
            </Table>
        )
    }
}

export default VTable