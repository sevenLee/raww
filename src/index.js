import _Table from './Table';
import _VTable from './VTable';
export {_Table as Table};
export {_VTable as VTable};