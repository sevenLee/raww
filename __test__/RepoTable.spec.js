import React from 'react'
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import {Table} from '../src'


describe('Raw Table Test', () => {
    let output = null

    beforeEach(() => {
        output = shallow(
            <Table datas={[]} />
        )
    })

    it('should render empty table correctly', () => {
        expect(shallowToJson(output)).toMatchSnapshot();
    });

    it('should render a table dom', () => {
        expect(output.find('.table-wrapper')).toHaveLength(1);
    });

    it('should render description correctly', () => {
        const output = shallow(
            <Table datas={[]} descriptionComp={<span>has description</span>}/>
        );
        expect(shallowToJson(output)).toMatchSnapshot();
        expect(output.find('.table-description')).toHaveLength(1);
        expect(output.find('.table-description').html()).toEqual('<div class="table-description"><span>has description</span></div>');
    });

    it('should render children (tfoot) correctly', () => {
        const outputWithChildren = shallow(
            <Table datas={[]}>
                <tfoot className="table-tfoot"><tr><td>has tfoot</td></tr></tfoot>
            </Table>
        );
        expect(shallowToJson(outputWithChildren)).toMatchSnapshot();
        expect(outputWithChildren.find('.table-tfoot')).toHaveLength(1);
        expect(outputWithChildren.find('.table-tfoot').html()).toEqual('<tfoot class="table-tfoot"><tr><td>has tfoot</td></tr></tfoot>');
    });



    describe('Raw Table render mock data',() => {
        let output = null
        let datas = [{playerId: 'playerId1', partner: 'a1'}, {playerId: 'playerId2', partner: 'a2'}]
        let colMap = {
            // firstColIndex: {index: 0, intlName: 'Index_Number',},
            playerId: {index:0, headerContent: () => 'Player ID'},
            partner: {index:1, headerContent: () => 'Partner'},
        }

        beforeEach(() => {
            output = shallow(<Table datas={datas} colMap={colMap}/>)
        })

        it('should render correctly with mock data', () => {
            expect(shallowToJson(output)).toMatchSnapshot();
        })

        it('should render 2 cols with data and column header', () => {
            expect(output.find('tbody > tr')).toHaveLength(2);
            expect(output.find('tbody > tr').at(0).childAt(0).text()).toEqual('playerId1');
            expect(output.find('tbody > tr').at(0).childAt(1).text()).toEqual('a1');
        })

        it('should render index column automatically when hasIndex true', () => {
            const colMapNew = {
                firstColIndex: {index: 0, headerContent: () => '#'},
                playerId: {index:1, headerContent: () => 'Player ID'},
                partner: {index:2, headerContent: () => 'Partner'},
            }

            const output = shallow(<Table datas={datas} colMap={colMapNew} config={{hasIndex: true}}/>)
            expect(shallowToJson(output)).toMatchSnapshot();
            expect(output.find('tbody > tr').at(0).children()).toHaveLength(3);
            expect(output.find('tbody > tr').at(0).childAt(0).text()).toEqual('1');
        })
    })
})