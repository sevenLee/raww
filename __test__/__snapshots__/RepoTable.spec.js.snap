// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Raw Table Test Raw Table render mock data should render correctly with mock data 1`] = `
<div
  className="table-wrapper"
>
  <Table
    bordered={false}
    bsClass="table"
    className="bo-table"
    condensed={false}
    hover={false}
    responsive={true}
    striped={false}
  >
    <thead>
      <tr>
        <th
          className="text-right"
          key="0"
          style={Object {}}
        >
          <span>
            Player ID
          </span>
        </th>
        <th
          className="text-right"
          key="1"
          style={Object {}}
        >
          <span>
            Partner
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className=""
        key="0"
        onClick={[Function]}
      >
        <td
          className="text-right"
          key="0"
          style={Object {}}
        >
          playerId1
        </td>
        <td
          className="text-right"
          key="1"
          style={Object {}}
        >
          a1
        </td>
      </tr>
      <tr
        className=""
        key="1"
        onClick={[Function]}
      >
        <td
          className="text-right"
          key="0"
          style={Object {}}
        >
          playerId2
        </td>
        <td
          className="text-right"
          key="1"
          style={Object {}}
        >
          a2
        </td>
      </tr>
    </tbody>
  </Table>
</div>
`;

exports[`Raw Table Test Raw Table render mock data should render index column automatically when hasIndex true 1`] = `
<div
  className="table-wrapper"
>
  <Table
    bordered={false}
    bsClass="table"
    className="bo-table"
    condensed={false}
    hover={false}
    responsive={true}
    striped={false}
  >
    <thead>
      <tr>
        <th
          className="text-right"
          key="0"
          style={
            Object {
              "width": 45,
            }
          }
        >
          <span>
            #
          </span>
        </th>
        <th
          className="text-right"
          key="1"
          style={Object {}}
        >
          <span>
            Player ID
          </span>
        </th>
        <th
          className="text-right"
          key="2"
          style={Object {}}
        >
          <span>
            Partner
          </span>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        className=""
        key="1"
        onClick={[Function]}
      >
        <td
          className=" text-index"
          key="0"
        >
          1
        </td>
        <td
          className="text-right"
          key="1"
          style={Object {}}
        >
          playerId1
        </td>
        <td
          className="text-right"
          key="2"
          style={Object {}}
        >
          a1
        </td>
      </tr>
      <tr
        className=""
        key="2"
        onClick={[Function]}
      >
        <td
          className=" text-index"
          key="0"
        >
          2
        </td>
        <td
          className="text-right"
          key="1"
          style={Object {}}
        >
          playerId2
        </td>
        <td
          className="text-right"
          key="2"
          style={Object {}}
        >
          a2
        </td>
      </tr>
    </tbody>
  </Table>
</div>
`;

exports[`Raw Table Test should render children (tfoot) correctly 1`] = `
<div
  className="table-wrapper"
>
  <Table
    bordered={false}
    bsClass="table"
    className="bo-table"
    condensed={false}
    hover={false}
    responsive={true}
    striped={false}
  >
    <thead>
      <tr />
    </thead>
    <tbody />
    <tfoot
      className="table-tfoot"
    >
      <tr>
        <td>
          has tfoot
        </td>
      </tr>
    </tfoot>
  </Table>
</div>
`;

exports[`Raw Table Test should render description correctly 1`] = `
<div
  className="table-wrapper"
>
  <div
    className="table-description"
  >
    <span>
      has description
    </span>
  </div>
  <Table
    bordered={false}
    bsClass="table"
    className="bo-table"
    condensed={false}
    hover={false}
    responsive={true}
    striped={false}
  >
    <thead>
      <tr />
    </thead>
    <tbody />
  </Table>
</div>
`;

exports[`Raw Table Test should render empty table correctly 1`] = `
<div
  className="table-wrapper"
>
  <Table
    bordered={false}
    bsClass="table"
    className="bo-table"
    condensed={false}
    hover={false}
    responsive={true}
    striped={false}
  >
    <thead>
      <tr />
    </thead>
    <tbody />
  </Table>
</div>
`;
